jsk_perception/src/histogram_of_oriented_gradients.cpp

```diff
@@ -1,99 +1,112 @@
+// Copyright (C) 2016 by Krishneel Chaudhary @ JSK Lab, The University
+// of Tokyo
 
 #include <jsk_perception/histogram_of_oriented_gradients.h>
 
-HOGFeatureDescriptor::HOGFeatureDescriptor() {
-   
+HOGFeatureDescriptor::HOGFeatureDescriptor(
+    const int cell_size, const int block_per_cell,
+    const int n_bins, const float angle, const int num_threads) :
+    CELL(cell_size),
+    BLOCK(block_per_cell),
+    ANGLE(angle),
+    N_BINS(n_bins),
+    num_threads_(num_threads_) {
+    this->BINS_ANGLE = this->ANGLE / this->N_BINS;
 }
 
-void HOGFeatureDescriptor::bilinearBinVoting(
+void HOGFeatureDescriptor::histogramBinVoting(
     const float &angle, int &lower_index, int &higher_index) {
     float nearest_lower = FLT_MAX;
     float nearest_higher = FLT_MAX;
     lower_index = 0;
     higher_index = 0;
     for (int i = BINS_ANGLE/2; i < ANGLE; i += BINS_ANGLE) {
-       float distance = abs(angle - i);
+	float distance = abs(angle - i);
         if (i < angle) {
             if (distance < nearest_lower) {
                 nearest_lower = distance;
                 lower_index = i;
             }
-        } else {
-            if (distance < nearest_higher) {
-                nearest_higher = distance;
-                higher_index = i;
-            }
         }
+	/*
+	else {
+	    if (distance < nearest_higher) {
+		nearest_higher = distance;
+		higher_index = i;
+	    }
+	}
+	*/
     }
 }
 
 void HOGFeatureDescriptor::imageGradient(
     const cv::Mat &image, cv::Mat &hog_bins) {
     cv::Mat xsobel;
     cv::Mat ysobel;
-    cv::Sobel(image, xsobel, CV_32F, 1, 0, 3);
-    cv::Sobel(image, ysobel, CV_32F, 0, 1, 3);
+    cv::Sobel(image, xsobel, CV_32F, 1, 0, 7);
+    cv::Sobel(image, ysobel, CV_32F, 0, 1, 7);
     cv::Mat Imag;
     cv::Mat Iang;
     cv::cartToPolar(xsobel, ysobel, Imag, Iang, true);
-    cv::add(Iang, cv::Scalar(180), Iang, Iang < 0);
-    cv::add(Iang, cv::Scalar(-180), Iang, Iang >= 180);
+    cv::add(Iang, cv::Scalar(ANGLE), Iang, Iang < 0);
```

Comment:
Basically, if the angle is less than 0, just add 180. Likewise if the angle is greater than 180, just subtract by 180.